File: src/bindings/wasm/in3.js
/*******************************************************************************
 * This file is part of the Incubed project.
 * Sources: https://github.com/slockit/in3-c
 * 
 * Copyright (C) 2018-2019 slock.it GmbH, Blockchains LLC
 * 
 * 
 * COMMERCIAL LICENSE USAGE
 * 
 * Licensees holding a valid commercial license may use this file in accordance 
 * with the commercial license agreement provided with the Software or, alternatively, 
 * in accordance with the terms contained in a written agreement between you and 
 * slock.it GmbH/Blockchains LLC. For licensing terms and conditions or further 
 * information please contact slock.it at in3@slock.it.
 * 	
 * Alternatively, this file may be used under the AGPL license as follows:
 *    
 * AGPL LICENSE USAGE
 * 
 * This program is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License as published by the Free Software 
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *  
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY 
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A 
 * PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.
 * [Permissions of this strong copyleft license are conditioned on making available 
 * complete source code of licensed works and modifications, which include larger 
 * works using a licensed work, under the same license. Copyright and license notices 
 * must be preserved. Contributors provide an express grant of patent rights.]
 * You should have received a copy of the GNU Affero General Public License along 
 * with this program. If not, see <https://www.gnu.org/licenses/>.
 *******************************************************************************/

// implement the transport and storage handlers
if (typeof fetch === 'function') {
    // for browsers
    in3w.in3_cache = {
        get: key => window.localStorage.getItem('in3.' + key),
        set: (key, value) => window.localStorage.setItem('in3.' + key, value)
    }
    in3w.transport = (url, payload) => fetch(url, {
        method: 'POST',
        mode: 'cors', // makes it possible to access them even from the filesystem.
        headers: { 'Content-Type': 'application/json' },
        body: payload
    }).then(res => {
        if (res.status < 200 || res.status >= 400) throw new Error("Error fetrching" + url + ":" + res.statusText)
        return res.text()
    })
}
else {
    // for nodejs
    fs = require('' + 'fs')
    try { fs.mkdirSync('.in3') } catch (x) { }
    in3w.in3_cache = {
        get(key) {
            try {
                return fs.readFileSync('.in3/' + key).toString('hex');
            } catch (x) { }
            return null
        },
        set(key, value) {
            fs.writeFileSync('.in3/' + key, Buffer.from(value, 'hex'))
        }
    }

    try {
        // if axios is available, we use it
        const axios = require('' + 'axios')
        in3w.transport = (url, payload) => axios.post(url, JSON.parse(payload), { headers: { 'Content-Type': 'application/json' } })
            .then(res => {
                if (res.status != 200) throw new Error("Invalid satus")
                return JSON.stringify(res.data)
            })
    } catch (xx) {
        // if not we use the raw http-implementation of nodejs
        in3w.transport = (url, payload) => new Promise((resolve, reject) => {
            try {
                const postData = payload;//JSON.stringify(payload);
                const m = require(url.startsWith('https') ? 'https' : 'http')
                const req = m.request(url, {
                    method: 'POST',
                    headers: {
                        'Content-Type': 'application/json',
                        'Content-Length': Buffer.byteLength(postData)
                    },
                    body: payload // body data type must match "Content-Type" header
                }, (res) => {
                    if (res.statusCode < 200 || res.statusCode >= 400)
                        reject(new Error("Invalid Status (" + res.statusCode + ') from server'))
                    else {
                        res.setEncoding('utf8');
                        let result = ''
                        res.on('data', _ => result += _)
                        res.on('end', () => resolve(result))
                    }
                })
                req.on('error', (e) => reject(new Error(e.message)))
                req.write(postData);
                req.end();

            }
            catch (er) {
                console.error('...ERROR : ', er)
                throw er
            }
        })
    }
}

let response_counter = 1
const responses = {}
function add_response(s) {
    responses['' + response_counter] = s
    return response_counter++
}
function get_response(n) {
    const r = responses['' + n]
    delete responses['' + n]
    return r
}

// signer-delegate
in3w.sign_js = async (clientPtr, type, message, account) => {
    const c = clients['' + clientPtr]
    if (!c) throw new Error('wrong client ptr')
    if (!c.signer) throw new Error('no signer set to handle signing')
    if (!(await c.signer.hasAccount(account))) throw new Error('unknown account ' + account)
    return await c.signer.sign(message, account, type)
}


const clients = {}



// create a flag ndicating when the wasm was succesfully loaded.
let _in3_listeners = []
in3w.onRuntimeInitialized = _ => {
    const o = _in3_listeners
    _in3_listeners = undefined
    o.forEach(_ => _(true))
}



function throwLastError() {
    const er = in3w.ccall('in3_last_error', 'string', [], []);
    if (er) throw new Error(er);
}
const aliases = { kovan: '0x2a', tobalaba: '0x44d', main: '0x1', ipfs: '0x7d0', mainnet: '0x1', goerli: '0x5' }

/**
 * The incubed client.
 */
class IN3 {

    // since loading the wasm is async, we always need to check whether the was was created before using it.
    async _ensure_ptr() {
        if (this.ptr) return
        if (_in3_listeners)
            await new Promise(r => _in3_listeners.push(r))
        this.ptr = in3w.ccall('in3_create', 'number', [], []);
        clients['' + this.ptr] = this
    }

    // here we are creating the instance lazy, when the first function is called.
    constructor(config) {
        this.config = config
        this.needsSetConfig = !!config
        this.ptr = 0;
        this.eth = new EthAPI(this)
    }

    /**
     * configures the client.
     */
    setConfig(conf) {
        if (conf) {
            if (conf.chainId) conf.chainId = aliases[conf.chainId] || conf.chainId
            this.config = { ...this.config, ...conf }
        }
        this.needsSetConfig = !this.ptr
        if (this.ptr) {
            const r = in3w.ccall('in3_config', 'number', ['number', 'string'], [this.ptr, JSON.stringify(this.config)]);
            if (r) throw new Error("Error setting the config : " + r);
        }
    }

    /**
       * sends one or a multiple requests.
       * if the request is a array the response will be a array as well.
       * If the callback is given it will be called with the response, if not a Promise will be returned.
       * This function supports callback so it can be used as a Provider for the web3.
       */
    send(request, callback) {
        const p = this.sendRequest(request)
        if (callback)
            p.then(_ => callback(null, _), err => callback(err, null))
        else
            return p
    }


    /**
     * sends a request and returns the response.
     */
    async sendRequest(rpc) {
        // ensure we have created the instance.
        if (!this.ptr) await this._ensure_ptr();
        if (this.needsSetConfig) this.setConfig()

        // create the context
        const r = in3w.ccall('in3_create_request', 'number', ['number', 'string'], [this.ptr, JSON.stringify(rpc)]);
        if (!r) throwLastError();

        // now send 
        // we add the pending request with pointer as key.
        function checkResponse(error) {
            // check if it was an error...
            const er = error || in3w.ccall('request_get_error', 'string', ['number'], [r])
            // if not we ask for the result.
            const res = er ? '' : in3w.ccall('request_get_result', 'string', ['number'], [r])

            // we always need to cleanup
            in3w.ccall('in3_free_request', 'void', ['number'], [r])

            // resolve or reject the promise.
            if (er) throw new Error(er)
            else {
                const r = JSON.parse(res)
                if (r) delete r.in3
                return r
            }
        }

        // send the request.
        return in3w.ccall('in3_send_request', 'void', ['number'], [r], { async: true }).then(checkResponse, checkResponse)
    }

    async sendRPC(method, params) {
        const res = await this.sendRequest({ method, params })
        if (res.error) throw new Error(res.error.message || res.error)
        return res.result
    }

    free() {
        if (this.ptr) {
            delete clients['' + this.ptr]
            in3w.ccall('in3_dispose', 'void', ['number'], [this.ptr])
            this.ptr = 0
        }
    }


}
// change the transport
IN3.setTransport = function (fn) {
    in3w.transport = fn
}

// change the transport
IN3.setStorage = function (fn) {
    in3w.in3_cache = fn
}

IN3.onInit = function (fn) {
    return new Promise((resolve, reject) => {
        const check = () => {
            try {
                resolve(fn())
            }
            catch (x) {
                reject(x)
            }
        }
        if (_in3_listeners)
            _in3_listeners.push(check)
        else
            check()
    })
}

// also support ES6-modules
IN3.default = IN3

// defined the export
if (typeof module !== "undefined")
    module.exports = IN3

File: java/src/in3/IN3.java
/*******************************************************************************
 * This file is part of the Incubed project.
 * Sources: https://github.com/blockchainsllc/in3
 *
 * Copyright (C) 2018-2020 slock.it GmbH, Blockchains LLC
 *
 *
 * COMMERCIAL LICENSE USAGE
 *
 * Licensees holding a valid commercial license may use this file in accordance
 * with the commercial license agreement provided with the Software or, alternatively,
 * in accordance with the terms contained in a written agreement between you and
 * slock.it GmbH/Blockchains LLC. For licensing terms and conditions or further
 * information please contact slock.it at in3@slock.it.
 *
 * Alternatively, this file may be used under the AGPL license as follows:
 *
 * AGPL LICENSE USAGE
 *
 * This program is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 * PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.
 * [Permissions of this strong copyleft license are conditioned on making available
 * complete source code of licensed works and modifications, which include larger
 * works using a licensed work, under the same license. Copyright and license notices
 * must be preserved. Contributors provide an express grant of patent rights.]
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <https://www.gnu.org/licenses/>.
 *******************************************************************************/

package in3;

import in3.config.ClientConfiguration;
import in3.eth1.SimpleWallet;
import in3.utils.Crypto;
import in3.utils.CurveType;
import in3.utils.JSON;
import in3.utils.PayloadType;
import in3.utils.SignatureType;
import in3.utils.Signer;
import in3.utils.StorageProvider;
import in3.utils.TransportException;

/**
 * This is the main class creating the incubed client. The client can then be
 * configured.
 *
 */
public class IN3 {
  private static final String CONFIG      = "in3_config";
  private static final String NODE_LIST   = "in3_nodeList";
  private static final String SIGN        = "in3_sign";
  private static final String CACHE_CLEAR = "in3_cacheClear";

  private static final String ENS_SUFFIX = ".ETH";

  static {
    Loader.loadLibrary();
  }

  private long            ptr;
  private StorageProvider provider;
  private Signer          signer;

  private static IN3Transport transport = new IN3DefaultTransport();
  private ClientConfiguration config;

  private IN3(long chainAlias) {
    ptr         = init(chainAlias);
    this.config = new ClientConfiguration(this.getDefaultConfig());
  }

  /**
   * creates a client with the default config.
   */
  public IN3() {
    ptr         = init(1);
    this.config = new ClientConfiguration(this.getDefaultConfig());
    this.config.markAsSynced();
  }

  /**
   * create a Incubed client using the chain-config. if chainId is
   * Chain.MULTICHAIN, the client can later be switched between different chains,
   * for all other chains, it will be initialized only with the chainspec for this
   * one chain (safes memory)
   */
  public static IN3 forChain(long chainId) {
    return new IN3(chainId);
  }

  protected void finalize() {
    free();
  }

  /**
   * sets config object in the client
   */
  private native void setConfig(String val);

  protected void setConfig(ClientConfiguration config) {
    this.config = config;
  }

  protected void applyConfig() {
    setConfig(config.toJSON());
    config.markAsSynced();
  }

  /**
   * returns the current configuration. any changes to the configuration will be
   * applied witth the next request.
   */
  public ClientConfiguration getConfig() {
    return config;
  }

  /**
   * sets the signer or wallet.
   */
  public void setSigner(Signer signer) {
    this.signer = signer;
  }

  /**
   * returns the signer or wallet.
   */
  public Signer getSigner() {
    return signer;
  }

  /**
   * gets the ipfs-api
   */
  public in3.ipfs.API getIpfs() {
    return new in3.ipfs.API(this);
  }

  /**
   * gets the zksync-api
   */
  public in3.zksync.API getZksync() {
    return new in3.zksync.API(this);
  }

  /**
   * gets the btc-api
   */
  public in3.btc.API getBtcAPI() {
    return new in3.btc.API(this);
  }

  /**
   * gets the ethereum-api
   */
  public in3.eth1.API getEth1API() {
    return new in3.eth1.API(this);
  }

  /**
   * gets the utils/crypto-api
   */
  public Crypto getCrypto() {
    return new Crypto(this);
  }

  /**
   * provides the ability to cache content like nodelists, contract codes and
   * validatorlists
   */
  public void setStorageProvider(StorageProvider val) {
    provider = val;
    initcache();
  }

  /**
   * provides the ability to cache content
   */
  public StorageProvider getStorageProvider() {
    return provider;
  }

  /**
   * sets The transport interface. This allows to fetch the result of the incubed
   * in a different way.
   */
  public void setTransport(IN3Transport newTransport) {
    IN3.transport = newTransport;
  }

  /**
   * returns the current transport implementation.
   */
  public IN3Transport getTransport() {
    return IN3.transport;
  }

  /**
   * servers to filter for the given chain. The chain-id based on EIP-155.
   */
  public native long getChainId();

  /**
   * sets the chain to be used. The chain-id based on EIP-155.
   */
  public native void setChainId(long val);

  /**
   * send a request. The request must a valid json-string with method and params
   */
  public String send(String request) {
    if (!config.isSynced()) {
      this.applyConfig();
    }
    return sendinternal(request);
  }

  private native String sendinternal(String request);

  /**
   * send a request but returns a object like array or map with the parsed
   * response. The request must a valid json-string with method and params
   */
  public Object sendobject(String request) {
    if (!config.isSynced()) {
      this.applyConfig();
    }
    return sendobjectinternal(request);
  }

  private native Object sendobjectinternal(String request);

  private String toRPC(String method, Object[] params) {
    StringBuilder p = new StringBuilder();
    for (int i = 0; i < params.length; i++) {
      if (p.length() > 0)
        p.append(",");
      if (params[i] == null)
        p.append("null");
      else if (params[i] instanceof byte[]) {
        byte[] b = (byte[]) params[i];
        p.append("\"0x");
        for (int j = 0; j < b.length; j++) {
          p.append(Character.forDigit((b[j] >> 4) & 0xF, 16));
          p.append(Character.forDigit(b[j] & 0xF, 16));
        }
        p.append("\"");
      }
      else if (params[i] instanceof String) {
        String s = (String) params[i];
        if (s.charAt(0) == '{' || s.equals("true") || s.equals("false"))
          p.append(s);
        else
          p.append("\"" + s + "\"");
      }
      else
        p.append(JSON.toJson(params[i]));
    }
    return "{\"method\":\"" + method + "\", \"params\":[" + p + "]}";
  }

  private String toRPC(String method, Object[] params, IN3Props props) {
    String p = "";
    for (int i = 0; i < params.length; i++) {
      if (p.length() > 0)
        p += ",";
      if (params[i] == null)
        p += "null";
      else if (params[i] instanceof String) {
        String s = (String) params[i];
        if (s.charAt(0) == '{' || s.equals("true") || s.equals("false"))
          p += s;
        else
          p += "\"" + s + "\"";
      }
      else
        p += JSON.toJson(params[i]);
    }

    return "{\"in3\":" + props.toJSON() + ", \"method\":\"" + method + "\", \"params\":[" + p + "]}";
  }

  /**
   * send a RPC request by only passing the method and params. It will create the
   * raw request from it and return the result.
   */
  public String sendRPC(String method, Object[] params) {
    return this.send(toRPC(method, params));
  }

  private Object sendObjectRPC(String method, Object[] params, IN3Props props) {
    return this.sendobject(toRPC(method, params, props));
  }

  public Object sendRPCasObject(String method, Object[] params, boolean useEnsResolver) {
    Object[] resolvedParams = useEnsResolver ? handleEns(params) : params;
    return this.sendobject(toRPC(method, resolvedParams));
  }

  /**
   * send a RPC request by only passing the method and params. It will create the
   * raw request from it and return the result.
   */
  public Object sendRPCasObject(String method, Object[] params) {
    return sendRPCasObject(method, params, true);
  }

  /**
   * internal function to handle the internal requests
   */
  static byte[][] sendRequest(String method, String[] urls, byte[] payload, String[] headers) throws TransportException {
    return IN3.transport.handle(method, urls, payload, headers);
  }

  private native void free();

  private native long init(long chainId);

  private native void initcache();

  private native Object getDefaultConfig();

  /**
   * returns the current incubed version.
   */
  public static native String getVersion();

  /**
   * clears the cache.
   */
  public boolean cacheClear() {
    return (boolean) sendRPCasObject(CACHE_CLEAR, new Object[] {});
  }

  /**
   * restrieves the node list
   */
  public IN3Node[] nodeList() {
    return nodeList(new String[] {});
  }

  /**
   * restrieves the node list
   */
  protected IN3Node[] nodeList(String[] signerNodeAddresses) {
    NodeList nl;
    if (signerNodeAddresses != null && signerNodeAddresses.length > 0) {
      IN3Props props = new IN3Props();
      props.setSignerNodes(signerNodeAddresses);
      nl = NodeList.asNodeList(sendObjectRPC(NODE_LIST, new Object[] {}, props));
    }

    nl = NodeList.asNodeList(sendRPCasObject(NODE_LIST, new Object[] {}));
    return nl.getNodes();
  }

  /**
   * request for a signature of an already verified hash.
   */
  public SignedBlockHash[] sign(BlockID[] blocks, String[] dataNodeAdresses) {
    if (dataNodeAdresses != null && dataNodeAdresses.length > 0) {
      IN3Props props = new IN3Props();
      props.setDataNodes(dataNodeAdresses);
      return SignedBlockHash.asSignedBlockHashs(sendObjectRPC(SIGN, new Object[] {blocks}, props));
    }

    return SignedBlockHash.asSignedBlockHashs(sendRPCasObject(SIGN, new Object[] {blocks}));
  }

  protected Object[] handleEns(Object[] params) {
    Object[] result = params.clone();
    for (int i = 0; i < result.length; i++) {
      if (result[i] != null && result[i].toString().toUpperCase().endsWith(ENS_SUFFIX)) {
        result[i] = (Object) getEth1API().ens(result[i].toString());
      }
    }

    return result;
  }

  // Test it
  public static void main(String[] args) {
    IN3          in3      = new IN3(Chain.GOERLI);
    SimpleWallet wallet   = new SimpleWallet();
    String       pk       = "0x889dbed9450f7a4b68e0732ccb7cd016dab158e6946d16158f2736fda1143ca6";
    String       signerId = wallet.addRawKey(pk);
    byte[] res            = wallet.sign("1e194c68360307cfb715bf17878791ad1ced8da7d2e5f42b691074c577f41eac",
                                        signerId, SignatureType.ethSign, PayloadType.ethTx, CurveType.ecdsa, null);

    System.out.println(res);
  }
}